### PR TITLE
feat(2.0): Optimized background sync

### PR DESCRIPTION
**Pull request by marc2332**
~Depends on https://github.com/iotaledger/iota-sdk/pull/1793 ( sorry, it was easier for me to test it, this way)~

# Description of change

- Remove unnecessary sleeps
- Remove an unnecessary extra async executor
- Use Tokio's `tokio::sync::watch` instead of AtomicUsize so it doesn't have to sleep and check the value, instead it uses [`wait_for`](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.wait_for)

Note: I haven't tried running this in WASM, but I suppose it should work just fine.

tdlr: More efficient and faster way to start/run/stop the background sync.

## Type of change

- Enhancement

## How the change has been tested

I tested it on Firefly, but this should be easy to check
- create a wallet
-  call start background sync
- check the logs to see if they are being properly generated
- call stop background sync
- check that the logs ended with the proper messages

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by Thoralf-M — DISMISSED** [review]

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/src/wallet/core/operations/background_syncing.rs [hunk 1]
```diff
@@ -1,18 +1,25 @@
...
-    wallet::{operations::syncing::SyncOptions, Wallet},
+    wallet::{operations::syncing::SyncOptions, task, Wallet},
 };
 
 /// The default interval for background syncing
 pub(crate) const DEFAULT_BACKGROUNDSYNCING_INTERVAL: Duration = Duration::from_secs(7);
 
+#[derive(Clone, PartialEq, Debug)]
+pub(crate) enum BackgroundSyncStatus {
+    NotRunning,
```
Comment (DaughterOfMars) [inline sdk/src/wallet/core/operations/background_syncing.rs]: Let's just make this `Stopped` I think

File: sdk/src/wallet/core/operations/background_syncing.rs [hunk 2]
```diff
@@ -25,69 +32,82 @@ where
...
+                log::debug!("[background_syncing]: syncing wallet");
+
+                if let Err(err) = wallet.sync(options.clone()).await {
+                    log::debug!("[background_syncing] error: {}", err)
                 }
-                wallet.background_syncing_status.store(0, Ordering::Relaxed);
-                log::debug!("[background_syncing]: stopped");
-            });
+
+                let seconds = interval.unwrap_or(DEFAULT_BACKGROUNDSYNCING_INTERVAL);
```
Comment (DaughterOfMars) [inline sdk/src/wallet/core/operations/background_syncing.rs]: move this up so we only do it once?

File: sdk/src/wallet/core/operations/background_syncing.rs [hunk 3]
```diff
@@ -49,16 +49,16 @@ where
...
 
         let wallet = self.clone();
+        let interval_seconds = interval.unwrap_or(DEFAULT_BACKGROUNDSYNCING_INTERVAL);
 
         task::spawn(async move {
             loop {
                 log::debug!("[background_syncing]: syncing wallet");
 
-                if let Err(err) = wallet.sync(options.clone()).await {
+                if let Err(err) = wallet.sync(interval_seconds.clone()).await {
```
Comment (DaughterOfMars) [inline sdk/src/wallet/core/operations/background_syncing.rs]: was this on purpose? shouldn't it still be `options`?

Comment (marc2332) [reply sdk/src/wallet/core/operations/background_syncing.rs]: Man... my vscodes crashed alongside my terminal (I blame rust-analyzer) and when I opened it again I went too fast on editing... :rofl: Anyway, just fixed it in the last commit  😅

**Review by Thoralf-M — DISMISSED** [review]

**Review by DaughterOfMars — DISMISSED** [review]

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Please rebase

**Review by DaughterOfMars — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Comment by thibault-martinez** [comment]
Thanks!

**Pull request merged**